### ISSUE TITLE
Expired AWS Token not retrieved again

### ISSUE THREAD

**Issue by rbobrovnikov** [kahing/goofys#503]
When credentials configured via AWS_SHARED_CREDENTIALS_FILE and mounted to the goofys container (in K8S) it successfully authenticates. After token expiration it updates in container but  goofys won't re-read it and I get "ExpiredToken: The provided token has expired" permanently.
Container restart solve this until token expired again.

In `github.com/aws/aws-sdk-go/aws/credentials/credentials.go` said:

> The Credentials is the primary method of getting access to and managing
credentials Values. Using dependency injection retrieval of the credential
values is handled by a object which satisfies the Provider interface.
By default the Credentials.Get() will cache the successful result of a
Provider's Retrieve() until Provider.IsExpired() returns true. At which
point Credentials will call Provider's Retrieve() to get new credential Value.
The Provider is responsible for determining when credentials Value have expired.
It is also important to note that Credentials will always call Retrieve the
first time Credentials.Get() is called.

But it seems doesn't work as expected.

Does goofy use `github.com/aws/aws-sdk-go/aws/credentials/shared_credentials_provider.go`?

**Comment by zhiburt** [comment]
Hey @rbobrovnikov. I have been chasing around with similar issue for a while. 

Yes goofys use this one, but as I can say `credentials.SharedCredentialsProvider` does not provide this option of rereading the config. Even though if you take a look at the source file you pointed out there's a comment which may perplex 
> // A SharedCredentialsProvider retrieves credentials from the current user's home
   // directory, and keeps track if those credentials are expired.

I hadn't come up with any ideas how to make it done. Eventually, I have created a wrapper of `credentials.Credentials` which contains the foregoing provider and simply call `credentials.Credentials.Expire` to force retrieving the credentials one more time. This method was called in consideration to expiration (estimated constant when the file can be updated).

And this model must seem pretty fine. At least for me at still does but I encountered the issue with data race when the binaries were deployed (#498) (I checked it after they'd been stroked).

It's an extordinary t

**Comment by zhiburt** [comment]
Hi @rbobrovnikov. I have gone through similar issue.

Yes goofys use this one, but as I can say `credentials.SharedCredentialsProvider` does not provide this option of rereading the config. Though if you take a look at the source file you pointed out there's a comment which may perplex

>    // A SharedCredentialsProvider retrieves credentials from the current user's home
>    // directory, and keeps track if those credentials are expired.

I hadn't come up with any ideas how to make it done. Eventually, I have created a wrapper of `credentials.Credentials` which contains the foregoing provider and simply call `credentials.Credentials.Expire` to force retrieving the credentials one more time. This method is called in consideration to expiration (estimated constant when the file could be updated).

And this model must seem pretty fine. At least for me at still does but I encountered the issue with data race when the binaries were deployed (#498). Precisely I decided to check it only when something extraordinary took place that is a bunch of instances were smashed out.

So until now I am not versed to put a statement that it's a utterly safe solution for a such type of issue. But I do can to say what I did after.

There was simply changed the strategy from rereading file to reading socket beside which another service obtain it by pooling. It's definitely got harder when the idea is the same. And actually data race issue is remains even with this one. But at first solution I have 5 crushes on 4 instances for 2 days, at the second one is up ideally for a long (nevertheless it may be just an coincidence).

If someone has resolved this issue please, let me know.

**Comment by kahing** [comment]
are you guys "expiring" credentials by generating a new `~/.aws/credentials`? I think maybe the proper way is to do something like https://docs.aws.amazon.com/cli/latest/topic/config-vars.html#sourcing-credentials-from-external-processes . This may require adding that support to aws-sdk-go

**Comment by zhiburt** [comment]
I tried exactly this approach.
The feature you have pointed out had been never heard of by me. As I see it could resolve it.

Surprisingly I have found out an associated [aws-sdk-go/#2217 PR](https://github.com/aws/aws-sdk-go/pull/2217) which is merged. And then I briefly cast an yeas over [the code](https://github.com/aws/aws-sdk-go/blob/87f0aa4be0153f21470e7a9802ebfb3403faa9ce/aws/credentials/processcreds/provider.go#L154). I suppose it provide it profoundly.

I wonder If it is authentic would you like have this feature provided by PR?:)

**Comment by kahing** [comment]
definitely. I think ideally all the user needs to do is specify `--profile`, and if it's configured for the default profile, nothing needs to be done. I imagine most of the work should be done in aws-sdk-go and probably little changes to goofys.

**Comment by zhiburt** [comment]
Apparently there's no need to change code. I have checked and it works well. Just pass the `--profile` option if the profile has `credential_process` field it will be used by aws-sdk-go.

I used this straightforward script as external process.

```bash
cat << EOF
{
  "Version": 1,
  "AccessKeyId": "your access key",
  "SecretAccessKey": "your secret access key"
}
EOF
```

Thanks for the comments.

**Comment by rbobrovnikov** [comment]
Actually I fixed this behaviour immediately after opening this issue simply with `credentials.NewSharedCredentials`.
In my case I use temporary aws token with limited lifetime and in my case it's more secure and convenient than pass static credentials to container.
If you don't mind I can make PR.

**Comment by kahing** [comment]
@rbobrovnikov what are you trying to change with your PR? As @zhiburt noted this already works without code change

**Comment by rbobrovnikov** [comment]
I have no permissions to send PR. Here's my tiny fix. It solve the case described in first post.

```
--- a/api/common/conf_s3.go
+++ b/api/common/conf_s3.go
@@ -35,6 +35,7 @@ type S3Config struct {
        RoleExternalId  string
        RoleSessionName string
        StsEndpoint     string
+        CredFilename    string

        RequesterPays bool
        Region        string
@@ -85,6 +86,7 @@ func (c *S3Config) ToAwsConfig(flags *FlagStorage) (*aws.Config, error) {
                if c.AccessKey != "" {
                        c.Credentials = credentials.NewStaticCredentials(c.AccessKey, c.SecretKey, "")
                }
+                c.Credentials = credentials.NewSharedCredentials(c.CredFilename, c.Profile)
        }
        if flags.Endpoint != "" {
                awsConfig.Endpoint = &flags.Endpoint
```
